### DISCUSSION TITLE
Getting data via direct access into Power BI using an On-Premise Gateway

### DISCUSSION THREAD

**Discussion by dahmhenderson** [belyaevvo/ClickHouseConnector#6] [Q&A]
We have created a new Power BI On-Premise gateway server for ClickHouse testing. We installed the ClickHouse ODBC drivers and using PowerBI Desktop on the gateway machine we can get direct query to work after placing the custom connector in the "...Power BI Desktop\Custom Connectors" directory. 

We now want to publish the report to Power BI Online but are unable to get the custom connector to work. Following some other instructions from Microsoft we copied the custom connector into the "C:\Windows\ServiceProfiles\PBIEgwService\Documents\Power BI Desktop\Custom Connectors" directory and loaded it onto the gateway but when we try to connect to ClickHouseODBC on the PowerBI gateway server from PowerBI Online we get an error message that we don't have any access? Has anyone else gotten it to work?
In summary we want to run a report in PowerBI Online using a Direct Query connection back to our On-Premise ClickHouse database via an On-Premise PowerBI Gateway. A similar things works with Hadoop ;)

**Comment by belyaevvo** [comment]
Do you install clickhouse odbc drivers on the gateway server? Please note,
you need to restart gateway service after coping connector and odbc
installation

пт, 10 нояб. 2023 г., 18:10 Stephen Henderson ***@***.***>:

**Reply by dahmhenderson** [reply]
Hello, thanks for your response. Yes I did install the drivers on the gateway and restart the service. 
The issue we have is that to get the Direct Query to work on PowerBI Desktop we need to use the ClickHouseConnectotor (Beta) locally as this is the only one which allows direct access. Even this needs renewed authentication and URL when configuring. The moment we publish the report to Power BI online the connection string cannot be added to the gateway as it says that we have no access to the ClickHouse database. 
Simply what we are looking for is a way to get Direct Query to an on-premise ClickHouse database from PowerBI Online using the PowerBI Gateway.

**Comment by belyaevvo** [comment]
I think this connector supports PowerBI on-premises gateway. As Microsoft's documentation says, the only thing that needs to be implemented compared to PowerBI Desktop is "For custom connectors to work with the on-premises data gateway, they need to implement the TestConnection section in the custom connector code." And I implemented it.

Please note that this document contains some limitations and considerations:
https://learn.microsoft.com/en-us/power-bi/connect-data/service-gateway-custom-connectors

As I remembered, I'd test connector in Gateway Scenario. So if you have any errors, please send me the error message you received.
I'll try to repeat Gateway test on Personal Gateway.

**Comment by dahmhenderson** [comment]
Hello, the error message I get is attached but I don't think it is actually telling the whole truth as the message never changes even if I use an incorrect URL or IP address. I have tried to look at the log files on the gateway but have not found anything. According to Google that is not unusual as even with the extended logs not everything is visible.
![ClickHouseError](https://github.com/belyaevvo/ClickHouseConnector/assets/55083296/684696fb-7483-438c-91d4-6550c76b59f4)
What I don't understand is why I need to type in the database information, username and password in to the connection config online as what we normally do is to choose anonymous, as the credentials etc. are already typed into the ODBC drive on the gateway?

**Comment by dahmhenderson** [comment]
Hello, 

the error message I get is attached but I don't think it is actually telling the whole truth as the message never changes even if I use an incorrect URL or IP address. I have tried to look at the log files on the gateway but have not found anything. According to Google that is not unusual as even with the extended logs not everything is visible.



What I don't understand is why I need to type in the database information, username and password in to the connection config online as what we normally do is to choose anonymous, as the credentials etc. are already typed into the ODBC drive on the gateway?

I have just reinstalled the Gateway and get the same issue, I can choose ClickHouseODBC as a Connection Type so the injection of the driver gets replicated online but saving gives the error message above.

￼
Any help you can give would be gratefully appreciated?

Regards,

Stephen Henderson
